File: app/components/folio/users/email_login/confirmation_component.js
window.Folio.Stimulus.register('f-users-email-login-confirmation', class extends window.Stimulus.Controller {
  static targets = ['status', 'retry']
  static values = { url: String, connectionError: String }

  connect () {
    this.token = window.location.hash ? window.location.hash.slice(1) : null
    if (window.location.hash) window.history.replaceState(null, '', window.location.pathname + window.location.search)
    this.prepare()
  }

  disconnect () {
    this.abortController?.abort()
    this.token = null
  }

  async prepare () {
    if (this.loading) return
    this.loading = true
    this.retryTarget.hidden = true
    this.abortController = new window.AbortController()
    try {
      const body = this.token === null ? {} : { email_login_token: this.token }
      const response = await window.Folio.Api.apiPost(this.urlValue, body, this.abortController.signal)
      this.token = null
      this.element.outerHTML = response.data
    } catch (error) {
      if (error.name !== 'AbortError') {
        this.statusTarget.textContent = this.connectionErrorValue
        this.retryTarget.hidden = false
      }
    } finally {
      this.loading = false
    }
  }
})

File: app/components/folio/users/email_login/confirmation_component.rb
# frozen_string_literal: true

class Folio::Users::EmailLogin::ConfirmationComponent < Folio::ApplicationComponent
  def initialize(challenge: nil, approved: false, invalid: false, preparing: false)
    @challenge = challenge
    @approved = approved
    @invalid = invalid
    @preparing = preparing
  end

  private
    def data
      return {} unless @preparing

      stimulus_controller("f-users-email-login-confirmation",
                          values: { url: controller.main_app.prepare_user_email_login_path(format: :json),
                                    connection_error: t("folio.users.email_login.prepare_error") })
    end

    def retry_data
      stimulus_merge_data(stimulus_target("retry"), stimulus_action(click: "prepare"))
    end
end
